File: code/python/algorithm_thinking_article-12.py
"""Commonly used bit operations"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().singleNumber(*args, **kwargs)


class _Solution:
    def singleNumber(self, nums: List[int]) -> int:
        res = 0
        for n in nums:
            res ^= n
        return res

File: code/java/algorithm_thinking_article-12.java
public class Algorithm_thinking_article_12 {
    /**
     * Commonly used bit operations
     */
    public Object solve(Object... args) {
        if (args == null || args.length != 1) {
            throw new IllegalArgumentException("Expected: (int[] nums)");
        }

        Solution solution = new Solution();
        return solution.singleNumber((int[]) args[0]);
    }

    private static class Solution {

        public int singleNumber(int[] nums) {
            int res = 0;
            for (int n : nums) {
                res ^= n;
            }
            return res;
        }
    }
}
